simSPI/linear_simulator/dataset_generator.py

```diff
@@ -0,0 +1,139 @@
+"""runs dataset generator and saves it."""
+import os
+
+import mrcfile
+import pandas as pd
+import starfile
+import torch
+
+from .modules import simulator
+from .params_utils import ParamsFactory
+from .starfile_utils import return_names, starfile_data
+
+"""Module to generate and save dataset (including metadata) in the output directory."""
+
+
+class dataset_gen(torch.nn.Module):
+    """class to generate and save dataset (including metadata) in the output directory.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the dataset generation and simulator.
+
+    """
+
+    def __init__(self, config):
+        super(dataset_gen, self).__init__()
+
+        self.config = config
+        self.simulator = simulator(config)  # instantiating the simulator
+        self.dataframe = []
+        self.init_dir()  # initializing the output folder
+        self.print_statements()
+        self.params_generator = ParamsFactory.get_params_generator(
+            config
+        )  # instantiating params generator
+
+    def run(self):
+        """Generate a chunk of projection and save it."""
+        for iterations in range(
+            self.config.datasetSize // self.config.chunks
+        ):  # TODO: last few images will be left
```

Comment:
you can use something like https://numpy.org/doc/stable/reference/generated/numpy.array_split.html to get pieces that are roughly the right size

or get indices and do [stard:end], check that end < len and do a final [start:]